EIPS/eip-3779.md

```diff
@@ -133,281 +89,196 @@ The remaining three states we can check at *validation time*:
+   * The number of items on the `return stack`MUST always be positive, and at most 1024.
+* The data stack MUST be consistently aligned:
+   * The number of items on the `data stack` between the current `stack pointer` and the `stack pointer` on entry to the most recent subroutine MUST be the same for each _execution_ of a _byte_code_.
 
-The `RJUMP`, `RJUMPI` and `RJUMPSUB`instructions take their destination as an immediate argument, so they are *static*.
+The `RJUMP`, `RJUMPI` and `RJUMPSUB` instructions take their destination as an immediate argument, which cannot change at runtime.  Having constant destinations for all jumps means that all jump destinations can be validated at initialization time, not runtime.  And requiring a consistently aligned`data stack` prevents stack underflow. It can also catch such errors as misaligned stacks due to irreducible control flows and calls to subroutines with the wrong number of arguments.
 
 Taken together, these rules allow for code to be validated by traversing the control-flow graph, in time and space linear in the size of the code, following each edge only once. 
 
-_Note: The definition of 'static' for `JUMP` and `JUMPI` is the bare minimum needed to implement subroutines.  Deeper analyses could be proposed that would validate a larger and probably more useful set of jumps, at the cost of more expensive (but still linear) validation._
-
-
 ## Rationale
 
-Demanding *static* destinations for all jumps means that all jump destinations can be validated at initialization time, not runtime.
+If we are to be able to validate contracts in linear time we have two real choices.
+
+1. Deprecate dynamic jumps. This is trivial.
+2. Constrain dynamic jumps. This requires static analysis.
 
-Bounding the stack pointers catches all `data stack` and non-recursive`return stack` overflows.
+Unfortunately the second choice is a nonstarter: we cannot always know whether a `JUMP` or `JUMPI` is a subroutine call in linear time, thus we must deprecate them. 
 
-Requiring a consistently aligned`data stack` prevents stack underflow. It can also catch such errors as misaligned stacks due to irreducible control flows and calls to subroutines with the wrong number of arguments.
+So far as exceptional halting, the states we validate and the runtime checks we require are the best that we candoe.  There are no doubt other useful safety properties that could be so validated, but we do not consider them here.
 
 ## Backwards Compatibility
 
-These changes affect the semantics of EVM code – the use of `JUMP`, `JUMPI`, and the stack are restricted, such that some *code* that would otherwise run correctly will nonetheless be invalid EVM *code*. 
+These changes affect the semantics of EVM code – the use of `JUMP`, `JUMPI`, and the stack are deprecated, such that some *code* that would otherwise run correctly will nonetheless be invalid EVM *code*.  In practice this will not be a problem, as these instruction are not valid in EOF code.
```

Comment:
```suggestion
These changes affect the semantics of EVM code – the use of `JUMP`, `JUMPI`, and the stack are deprecated, such that some *code* that would otherwise run correctly will nonetheless be invalid EVM *code*. In practice, this will not be a problem, as these instructions are not valid in EOF code.
```